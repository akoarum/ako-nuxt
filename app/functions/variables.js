export const VALIDATES = {
  EMPTY: {
    message: '入力してください。'
  },
  NUMBER: {
    regex: /^[0-9]+?$/,
    message: '数字で入力してください。'
  },
  EMAIL: {
    regex: /^[-a-zA-Z0-9!#$%&'*+\-/=?^_`.{|}~]+@[-a-zA-Z0-9]+(\.[-a-zA-Z0-9]+)*\.[a-zA-Z]{2,6}$/,
    message: 'メールアドレスの形式になっていません。'
  },
  TEL: {
    regex: /^[0-9]{2,4}-?[0-9]{2,4}-?[0-9]{4}$/,
    message: '電話番号の形式になっていません。'
  },
  URL: {
    regex: /^http(s):\/\/.+?\.+?$/,
    message: 'URLの形式になっていません。'
  }
}
