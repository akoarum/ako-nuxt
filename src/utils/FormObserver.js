export default class FormValidator {
  constructor(names = []) {
    this._items = this._createItems(names)
  }

  /**
   * @return {Boolean}
   */
  get btnDisabled() {
    for (const name in this._items) {
      if (this._items[name].error || !this._items[name].dirty) return true
    }
    return false
  }

  /**
   * itemオブジェクトの生成
   * @param {Array} names name配列
   */
  _createItems(names) {
    const items = {}

    names.forEach(item => {
      items[item] = {
        error: false,
        dirty: false
      }
    })

    return items
  }

  /**
   * itemの更新
   * @param {String}  name  name値
   * @param {String}  prop  更新するプロパティ名
   * @param {Boolean} value 値
   */
  updateItem(name, prop, value) {
    this._items[name][prop] = value
  }
}
