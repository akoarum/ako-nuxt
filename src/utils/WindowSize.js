export default class WindowSize {
  constructor(delay = 80) {
    this._width = 0
    this._height = 0
    this._delay = delay
    this._cullTime = null
    this._initialized = false
    this._handler = this._handleResize.bind(this)
  }

  get width() {
    return this._width
  }

  get height() {
    return this._height
  }

  get initialized() {
    return this._initialized
  }

  /**
   * 初期化メソッド
   */
  initialize() {
    if (this._initialized) return this
    this._updateState()
    window.addEventListener('resize', this._handler)
    this._initialized = true
    return this
  }

  /**
   * ステートの更新
   */
  _updateState() {
    this._width = window.innerWidth
    this._height = window.innerHeight
  }

  /**
   * リサイズハンドラ
   */
  _handleResize() {
    clearTimeout(this._cullTime)
    this._cullTime = setTimeout(() => {
      this._updateState()
    }, this._delay)
  }

  /**
   * 解放する
   */
  destroy() {
    if (!this._initialized) return this
    window.removeEventListener('resize', this._handler)
    this._initialized = false
    return this
  }
}
