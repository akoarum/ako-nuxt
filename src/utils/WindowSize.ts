export default class WindowSize {
  private _width: number = 0
  private _height: number = 0
  private delay: number = 80
  private cullTime = null
  private _initialized: boolean = false
  private handler = this.handleResize.bind(this)

  constructor(delay: number = 80) {
    this.delay = delay
  }

  public get width() {
    return this._width
  }

  public get height() {
    return this._height
  }

  public get initialized() {
    return this._initialized
  }

  /**
   * 初期化メソッド
   */
  public initialize() {
    if (this._initialized) return this
    this.updateState()
    window.addEventListener('resize', this.handler)
    this._initialized = true
    return this
  }

  /**
   * ステートの更新
   */
  public updateState() {
    this._width = window.innerWidth
    this._height = window.innerHeight
  }

  /**
   * リサイズハンドラ
   */
  private handleResize() {
    clearTimeout(this.cullTime)
    this.cullTime = setTimeout(() => {
      this.updateState()
    }, this.delay)
  }

  /**
   * 解放する
   */
  public destroy() {
    if (!this._initialized) return this
    window.removeEventListener('resize', this.handler)
    this._initialized = false
    return this
  }
}
