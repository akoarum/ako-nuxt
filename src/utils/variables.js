export const VALIDATES = {
  required: '入力してください。',
  selected: '選択してください。',
  file: 'ファイルをアップロードしてください。',
  tel: {
    regex: /^\d{2,4}-?\d{4}-?\d{4}$/,
    message: '電話番号は半角数字とハイフンで入力してください。'
  },
  email: {
    regex: /^[-a-zA-Z0-9!#$%&'*+\-/=?^_`.{|}~]+@[-a-zA-Z0-9]+(\.[-a-zA-Z0-9]+)*\.[a-zA-Z]{2,6}$/,
    message: 'メールアドレスを正しく入力してください。'
  }
}
