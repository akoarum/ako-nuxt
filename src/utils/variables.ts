import { ValidatePattern } from '~/types/models'

export const VALIDATE_PATTERNS: { [key: string]: ValidatePattern } = {
  required: {
    message: '入力してください。'
  },
  selected: {
    message: '選択してください。'
  },
  tel: {
    regex: /^\d{2,5}-?\d{1,4}-?\d{4}$/,
    message: '電話番号は半角数字とハイフンで入力してください。'
  },
  email: {
    regex: /^[-a-zA-Z0-9!#$%&'*+\-/=?^_`.{|}~]+@[-a-zA-Z0-9]+(\.[-a-zA-Z0-9]+)*\.[a-zA-Z]{2,6}$/,
    message: 'メールアドレスを正しく入力してください。'
  }
}
