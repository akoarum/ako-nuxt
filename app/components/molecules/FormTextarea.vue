<template>
  <div class="textarea">
    <VTextarea
      :id="id ? id : name"
      :name="name"
      :value="value"
      :placeholder="placeholder"
      :minlength="minlength"
      :maxlength="maxlength"
      :autocomplete="autocomplete"
      :disabled="disabled"
      :readonly="readonly"
      :required="required"
      :error="!!error"
      :aria-describedby="id ? `${id}-error` : `${name}-error`"
      @dirty="isDirtied = true"
      @input="inputHandler"
      class="textarea__body"
    />
    <VTexts
      v-show="error"
      :id="id ? `${id}-error` : `${name}-error`"
      role="alert"
      :text="error"
      :aria-live="error ? 'assertive' : 'off'"
      error
      class="textarea__error"
    />
  </div>
</template>

<script>
import { VALIDATES } from '~/functions/variables'
import VTextarea from '~/components/atoms/VTextarea'
import VTexts from '~/components/atoms/VTexts'

export default {
  components: { VTextarea, VTexts },
  props: {
    name: { type: String, required: true },
    value: { type: String, required: true },
    id: { type: String },
    placeholder: { type: String },
    minlength: { type: String },
    maxlength: { type: String },
    autocomplete: { type: Boolean },
    disabled: { type: Boolean },
    readonly: { type: Boolean },
    required: { type: Boolean }
  },
  data() {
    return {
      isDirtied: false,
      error: ''
    }
  },
  watch: {
    value() {
      this.error = this.validate()
    },
    isDirtied() {
      this.error = this.validate()
    },
    error() {
      this.$emit('error', { name: this.name, error: !!this.error })
    }
  },
  created() {},
  methods: {
    inputHandler(value) {
      this.$emit('input', value)
    },
    validate() {
      if (!this.isDirtied) {
        return ''
      }

      if (this.required && !this.value) {
        return VALIDATES.EMPTY.message
      }

      return ''
    }
  }
}
</script>

<style lang="scss" scoped>
.textarea {
  @media screen and (min-width: 768px) {
    &__error {
      margin-top: 4px;
    }
  }

  @media screen and (max-width: 767px) {
    &__error {
      margin-top: 3px;
    }
  }
}
</style>
