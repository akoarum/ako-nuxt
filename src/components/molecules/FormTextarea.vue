<template>
  <div class="formTextarea">
    <VTextarea
      v-model="model"
      v-bind="$attrs"
      :name="name"
      :required="required"
      :minlength="minlength"
      :maxlength="maxlength"
      :aria-invalid="String(!!error)"
      :aria-describedby="error ? `error-${name}` : false"
      class="formTextarea__textarea"
      @blur="updateDirty"
    />
    <VTexts
      v-if="error"
      :id="`error-${name}`"
      caution
      small
      class="formTextarea__error"
      role="alert"
      aria-live="assertive"
    >
      <span>{{ error }}</span>
    </VTexts>
  </div>
</template>

<script>
import { VALIDATES } from '~/utils/variables'
import { VTextarea, VTexts } from '~/components/atoms'

export default {
  components: { VTextarea, VTexts },
  inheritAttrs: false,
  props: {
    name: { type: String, required: true },
    value: { type: String, required: true },
    required: { type: Boolean },
    minlength: { type: Number },
    maxlength: { type: Number }
  },
  data() {
    return {
      error: '',
      isDirty: !!this.value || !this.required
    }
  },
  computed: {
    model: {
      set(value) {
        this.$emit('input', value)
        this.error = this.validate(value)
      },
      get() {
        return this.value
      }
    }
  },
  watch: {
    error(value) {
      this.$emit('error', { name: this.name, error: !!value })
    },
    isDirty() {
      this.$emit('dirty', this.name)
    }
  },
  methods: {
    validate(value) {
      if (!this.isDirty) return ''
      if (this.required && !value) return VALIDATES.required
      if (this.minlength && value.length < this.minlength) {
        return `${this.minlength}字以上で入力して下さい。`
      }
      if (this.maxlength && value.length > this.maxlength) {
        return `${this.maxlength}字以内で入力して下さい。`
      }
      return ''
    },
    updateDirty() {
      if (!this.value || this.isDirty) return false
      this.isDirty = true
      this.error = this.validate(this.value)
    }
  }
}
</script>

<style lang="scss" scoped>
.formTextarea__error {
  @include media-min {
    margin-top: 6px;
  }
  @include media-max {
    margin-top: 8px;
  }
}
</style>
