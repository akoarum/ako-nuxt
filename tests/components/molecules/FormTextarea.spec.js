import { shallowMount } from '@vue/test-utils'
import { VALIDATES } from '~/utils/variables'
import FormTextarea from '~/components/molecules/FormTextarea'

describe('FormTextarea', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(FormTextarea, {
      propsData: {
        name: 'test',
        value: '',
        required: true
      }
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })
    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('methods', () => {
    describe('validate', () => {
      it('requiredでvalueがない場合は空文字列を返す', () => {
        wrapper.setProps({ required: false })
        expect(wrapper.vm.validate('')).toBe('')
      })

      it('isDirtyでない場合、空文字列を返す', () => {
        expect(wrapper.vm.validate('')).toBe('')
      })

      it('isDirtyでrequired、valueが空の場合はREQUIREDエラーを返す', () => {
        wrapper.setData({ isDirty: true })
        expect(wrapper.vm.validate('')).toBe(VALIDATES.required)
      })

      it('isDirtyでminlengthよりvalueの文字数が少ない場合、エラーを返す', () => {
        wrapper.setProps({ minlength: 10 })
        wrapper.setData({ isDirty: true })
        expect(wrapper.vm.validate('a')).toBe(`10字以上で入力して下さい。`)
      })

      it('isDirtyでmaxlengthよりvalueの文字数が多い場合、エラーを返す', () => {
        wrapper.setProps({ maxlength: 10 })
        wrapper.setData({ isDirty: true })
        expect(wrapper.vm.validate('abcdefghijkl')).toBe(`10字以内で入力して下さい。`)
      })
    })

    describe('updateDirty', () => {
      it('valueが空でなくisDirtyでなければisDirtyをtrueにする', () => {
        const mockValidate = jest.fn()
        wrapper.setProps({ value: 'test' })
        wrapper.setMethods({ validate: mockValidate })
        wrapper.vm.updateDirty()
        expect(wrapper.vm.isDirty).toBe(true)
        expect(mockValidate).toHaveBeenCalled()
      })
    })
  })
})
