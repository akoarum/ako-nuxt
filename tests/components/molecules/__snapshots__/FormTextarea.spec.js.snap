// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTextarea initialized snapshot 1`] = `
<div
  class="formTextarea"
>
  <vtextarea-stub
    aria-invalid="false"
    class="formTextarea__textarea"
    name="test"
    required="true"
    value=""
  />
   
  <!---->
</div>
`;
